File: clients/php/Model/UnitOrder.php
<?php

namespace Model {
    require_once 'Model/ActionOrder.php';
    require_once 'Model/Vec2.php';
    require_once 'Stream.php';

    /**
     * Order for specific unit
     */
    class UnitOrder
    {
        /**
         * Target moving velocity
         */
        public \Model\Vec2 $targetVelocity;
        /**
         * Target view direction (vector length doesn't matter)
         */
        public \Model\Vec2 $targetDirection;
        /**
         * Order to perform an action, or None
         */
        public ?\Model\ActionOrder $action;
    
        function __construct(\Model\Vec2 $targetVelocity, \Model\Vec2 $targetDirection, ?\Model\ActionOrder $action)
        {
            $this->targetVelocity = $targetVelocity;
            $this->targetDirection = $targetDirection;
            $this->action = $action;
        }
    
        /**
         * Read UnitOrder from input stream
         */
        public static function readFrom(\InputStream $stream): UnitOrder
        {
            $targetVelocity = \Model\Vec2::readFrom($stream);
            $targetDirection = \Model\Vec2::readFrom($stream);
            if ($stream->readBool()) {
                $action = \Model\ActionOrder::readFrom($stream);
            } else {
                $action = NULL;
            }
            return new UnitOrder($targetVelocity, $targetDirection, $action);
        }
        
        /**
         * Write UnitOrder to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $this->targetVelocity->writeTo($stream);
            $this->targetDirection->writeTo($stream);
            if (is_null($this->action)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $this->action->writeTo($stream);
            }
        }
    }
}

File: clients/go/pkg/model/unit_order.go
package model

import (
	"aicup22/pkg/stream"
)

// Order for specific unit
type UnitOrder struct {
	// Target moving velocity
	TargetVelocity Vec2
	// Target view direction (vector length doesn't matter)
	TargetDirection Vec2
	// Order to perform an action, or None
	Action *ActionOrder
}

func NewUnitOrder(targetVelocity, targetDirection Vec2, action *ActionOrder) UnitOrder {
	return UnitOrder{
		TargetVelocity:  targetVelocity,
		TargetDirection: targetDirection,
		Action:          action,
	}
}

// Read UnitOrder from reader
func ReadUnitOrder() UnitOrder {
	targetVelocity := ReadVec2()
	targetDirection := ReadVec2()
	var action *ActionOrder
	if stream.Flow().ReadBool() {
		actionValue := ReadActionOrder()
		action = &actionValue
	} else {
		action = nil
	}
	return UnitOrder{
		TargetVelocity:  targetVelocity,
		TargetDirection: targetDirection,
		Action:          action,
	}
}

// Write UnitOrder to writer
func (sf UnitOrder) Write() {
	targetVelocity := sf.TargetVelocity
	targetVelocity.Write()
	targetDirection := sf.TargetDirection
	targetDirection.Write()
	action := sf.Action
	if action == nil {
		flow.WriteBool(false)
	} else {
		flow.WriteBool(true)
		actionValue := *action
		actionValue.Write()
	}
}

// Get string representation of UnitOrder
func (sf UnitOrder) String() string {
	strRes := "{ "
	strRes += "TargetVelocity: "
	targetVelocity := sf.TargetVelocity
	strRes += targetVelocity.String()
	strRes += ", "
	strRes += "TargetDirection: "
	targetDirection := sf.TargetDirection
	strRes += targetDirection.String()
	strRes += ", "
	strRes += "Action: "
	action := sf.Action
	if action == nil {
		strRes += "nil"
	} else {
		actionValue := *action
		strRes += actionValue.String()
	}
	strRes += " }"
	return strRes
}
